crowbar_framework/app/models/service_object.rb

```diff
@@ -1279,6 +1395,8 @@ def apply_role(role, inst, in_queue)
       end
 
       # Make sure the config role is on the nodes in this barclamp, otherwise remove it
+      # FIXME: All nodes is basically all new nodes, which should have an add/remove list,
+      # why do we need to add specifically the passed role?
```

Comment:
Ah, okay. It's two different roles :-) Let's say you add node1 to keystone-server and apply the keystone proposal, then the lines above make sure that in chef, node1 has the keystoner-server role. The lines below, on the other hand, make sure that node1 has the keystone-config-default role (which is the role matching this specific proposal, with the attributes that are set to what is in the proposal).

So you have two sets of roles:

- the roles like keystone-server that define what recipes will be run
- the roles like keystone-config-default that define what attributes are used to configure things properly during the chef run